README
NAME
    accessors - create accessor methods in caller's package.

SYNOPSIS
      package Foo;
      use accessors qw( foo bar baz );

      my $obj = bless {}, 'Foo';

      # generates chaining accessors
      # that you can set like this:
      $obj->foo( 'hello ' )
          ->bar( 'world' )
          ->baz( "!\n" );

      # you get the values by passing no params:
      print $obj->foo, $obj->bar, $obj->baz;

DESCRIPTION
    The accessors pragma lets you create simple accessors at compile-time.

    This saves you from writing them by hand, which tends to result in
    *cut-n-paste* errors and a mess of duplicated code. It can also help you
    reduce the ammount of unwanted *direct-variable access* that may creep
    into your codebase when you're feeling lazy. accessors was designed with
    laziness in mind.

    Method-chaining accessors are generated by default. Note that you can
    still use accessors::chained directly for reasons of backwards
    compatibility.

    See accessors::rw for accessors that always return the current value if
    you don't like method chaining.

GENERATED METHODS
    accessors will generate methods that return the current object on set:

      sub foo {
          my $self = shift;
          if (@_) { $self->{-foo} = shift; return $self; }
          else    { return $self->{-foo}; }
      }

    This way they can be *chained* together.

  Why prepend the dash?
    The dash ("-") is prepended to the property name for a few reasons:

    *   interoperability with Error.

    *   to make it difficult to accidentally access the property directly
        ala:

          use accessors qw( foo );
          $obj->{foo};  # prevents this by mistake
          $obj->foo;    # when you probably meant this

        (this might sound woolly, but it's easy enough to do).

    *   syntactic sugar (this *is* woolly :).

    You shouldn't care too much about how the property is stored anyway - if
    you do, you're likely trying to do something special (and should really
    consider writing the accessors out long hand), or it's simply a matter
    of preference in which case you can use accessors::rw, or sub-class this
    module.

PERFORMANCE
    There is little-to-no performace hit when using generated accessors; in
    fact there is usually a performance gain.

    *   typically *10-30% faster* than hard-coded accessors (like the above
        example).

    *   typically *1-15% slower* than *optimized* accessors (less readable).

    *   typically a *small* performance hit at startup (accessors are
        created at compile-time).

    *   uses the same anonymous sub to reduce memory consumption (sometimes
        by 80%).

    See the benchmark tests included with this distribution for more
    details.

MOTIVATION
    The main difference between the accessors pragma and other accessor
    generators is simplicity.

    *   interface

        use accessors qw( ... ) is as easy as it gets.

    *   a pragma

        it fits in nicely with the base pragma:

          use base      qw( Some::Class );
          use accessors qw( foo bar baz );

        and accessors get created at compile-time.

    *   no bells and whistles

        The module is extensible instead.

SUB-CLASSING
    If you prefer a different style of accessor or you need to do something
    more complicated, there's nothing to stop you from sub-classing. It
    should be pretty easy. Look through accessors::classic, accessors::ro,
    and accessors::rw to see how it's done.

CAVEATS
    Classes using blessed scalarrefs, arrayrefs, etc. are not supported for
    sake of simplicity. Only hashrefs are supported.

  Class Accessors Are Not Supported
    If you are accidentally calling the accessor as a class method:

      my $object = 'Foo'; # an accident!
      print $object->bla, "\n"; # does not die!

    Then (as of v1.02) this will produce an error ala:

      Can't use string ("Foo") as a HASH ref while "strict refs" in use
      at lib/accessors.pm line 72.

THANKS
    Thanks to Michael G. Schwern for indirectly inspiring this module, and
    for his feedback & suggestions.

    Also to Paul Makepeace and David Wright for showing me faster accessors,
    to chocolateboy & others for their contributions, the CPAN Testers for
    their bug reports, and to James Duncan and people on London.pm for their
    feedback.

AUTHORS
    Steve Purkis <spurkis@cpan.org>

    Contributions from:

      chocolateboy
      Slaven Rezić

SOURCE
    <https://github.com/spurkis/Perl-accessors>

SEE ALSO
    accessors::classic, accessors::chained

    Similar and related modules:

    base, fields, Class::Accessor, Moose, Class::Struct, Class::Methodmaker,
    Class::Generate, Class::Class, Class::Tangram, Object::Tiny



Changes
Revision history for 'accessors' Perl pragma
--------------------------------------------

All changes by Steve Purkis, unless otherwise noted.

1.02
    *	fixed bug: Prohibit "class accessors" RT #73138 [patch by Slaven Rezić]
	calling class accessors will now throw an exception.
    +	moved repo to github: git://github.com/spurkis/Perl-accessors.git
    +	fixed spelling errors in pod: RT #71663 [patch by Salvatore Bonaccorso]

1.01
    +	only run benchmark tests when $ENV{BENCHMARK_ACCESSORS} is set.
	(tests were failing randomly, and annoying lots of people)

1.00
    +	released as 1.0 to solidify the API & guarantee backwards compat
    +	fixed RT #29753: Suppress warnings in "make test"
	(it was a bug in the test code)
    +	included accessors::ro (read only) and accessors::rw (read/write)
	on chocolateboy's request.
    *	changed performance test to be more lenient:
	'set/get generated is < 15% slower than optimized (X %)'
	Threshold is now < 30% slower, because it was failing on some platforms
	(probably due to the amount of warnings generated by RT #29753, but
	 upped the threshold anyway).
    +	set more invalid accessor names [updated list from chocolateboy]


0.02
    +	made method-chaining accessors the default.
    *	created accessors::classic and accessors::chained, and killed the
	'use accessors qw( foo :style )' syntax.
    +	now compat with perl 5.6+ [reported by Michael Schwern]
    +	tests no longer noisy (no blib or benchmark diagnostics)
	[reported by Michael Schwern]
    +	using closures + anon sub instead of eval [reported by Michael Schwern]
    +	moved generation methods into simple subs & optimized (smaller pad to
	restore, results in faster, well factored code)
    +	added InvalidNames to catch things like AUTOLOAD & DESTROY
	[reported by James Duncan]
    +	s/cascading/chaining/ [reported by James Duncan]
    +	refactored test suite, added benchmarks for classic accessors.
    +	updated docs

0.01
    +	created Fri Sep 12 2003 after a conversation with Michael Schwern.


lib/accessors/ro.pm
=head1 NAME

accessors::ro - create 'classic' read-only accessor methods in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors::ro qw( foo bar baz );

  my $obj = bless { foo => 'read only? ' }, 'Foo';

  # values are read-only, so set is disabled:
  print "oh my!\n" if $obj->foo( "set?" ) eq 'read only? ';

  # if you really need to change the vars,
  # you must use direct-variable-access:
  $obj->{bar} = 'i need a drink ';
  $obj->{baz} = 'now';

  # always returns the current value:
  print $obj->foo, $obj->bar, $obj->baz, "!\n";

=cut

package accessors::ro;

use strict;
use warnings::register;
use base qw( accessors );

our $VERSION  = '1.00';
our $REVISION = (split(/ /, ' $Revision: 1.4 $ '))[2];

use constant style => 'ro';

sub create_accessor {
    my ($class, $accessor, $property) = @_;
    # get is slightly faster if we eval instead of using a closure + anon
    # sub, but the difference is marginal (~5%), and this uses less memory...
    no strict 'refs';
    *{$accessor} = sub { return $_[0]->{$property} };
}

1;

__END__

=head1 DESCRIPTION

The B<accessors::ro> pragma lets you create simple I<classic> read-only
accessors at compile-time.

The generated methods look like this:

  sub foo {
      my $self = shift;
      return $self->{foo};
  }

They I<always> return the current value, just like L<accessors::ro>.

=head1 PERFORMANCE

There is B<little-to-no performace hit> when using generated accessors; in
fact there is B<usually a performance gain>.

=over 4

=item *

typically I<5-15% faster> than hard-coded accessors (like the above example).

=item *

typically I<0-15% slower> than I<optimized> accessors (less readable).

=item *

typically a I<small> performance hit at startup (accessors are created at
compile-time).

=item *

uses the same anonymous sub to reduce memory consumption (sometimes by 80%).

=back

See the benchmark tests included with this distribution for more details.

=head1 CAVEATS

Classes using blessed scalarrefs, arrayrefs, etc. are not supported for sake
of simplicity.  Only hashrefs are supported.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>

=head1 SEE ALSO

L<accessors>,
L<accessors::rw>,
L<accessors::classic>,
L<accessors::chained>,
L<base>

=cut


lib/accessors/rw.pm
=head1 NAME

accessors::rw - create 'classic' read/write accessor methods in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors::rw qw( foo bar baz );

  my $obj = bless {}, 'Foo';

  # always return the current value, even on set:
  $obj->foo( 'hello ' ) if $obj->bar( 'world' ) eq 'world';

  print $obj->foo, $obj->bar, $obj->baz( "!\n" );

=cut

package accessors::rw;

use strict;
use warnings::register;
use base qw( accessors::classic );

our $VERSION  = '1.00';
our $REVISION = (split(/ /, ' $Revision: 1.3 $ '))[2];

use constant style => 'rw';

1;

__END__

=head1 DESCRIPTION

The B<accessors::rw> pragma lets you create simple I<classic> read/write
accessors at compile-time.  It is an alias for L<accessors::classic>.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>.

=head1 SEE ALSO

L<accessors>,
L<accessors::ro>,
L<accessors::classic>,
L<accessors::chained>,
L<base>

=cut


lib/accessors/chained.pm
=head1 NAME

accessors::chained - create method chaining accessors in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors::chained qw( foo bar baz );

  my $obj = bless {}, 'Foo';

  # generates chaining accessors:
  $obj->foo( 'hello ' )
      ->bar( 'world' )
      ->baz( "!\n" );

  print $obj->foo, $obj->bar, $obj->baz;

=cut

package accessors::chained;

use strict;
use warnings::register;
use base qw( accessors );

our $VERSION  = '1.00';
our $REVISION = (split(/ /, ' $Revision: 1.3 $ '))[2];

# inherit everything for now.

1;

__END__

=head1 DESCRIPTION

The B<accessors::chained> pragma lets you create simple method-chaining
accessors at compile-time.

This module exists for future backwards-compatability - if the default style
of accessor ever changes, method-chaining accessors will still be available
through this pragma.

See L<accessors> for documentation.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>

=head1 SEE ALSO

L<accessors>, L<accessors::classic>, L<base>

=cut


lib/accessors/classic.pm
=head1 NAME

accessors::classic - create 'classic' read/write accessor methods in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors::classic qw( foo bar baz );

  my $obj = bless {}, 'Foo';

  # always return the current value, even on set:
  $obj->foo( 'hello ' ) if $obj->bar( 'world' ) eq 'world';

  print $obj->foo, $obj->bar, $obj->baz( "!\n" );

=cut

package accessors::classic;

use strict;
use warnings::register;
use base qw( accessors );

our $VERSION  = '1.00';
our $REVISION = (split(/ /, ' $Revision: 1.5 $ '))[2];

use constant style => 'classic';

sub create_accessor {
    my ($class, $accessor, $property) = @_;
    # set/get is slightly faster if we eval instead of using a closure + anon
    # sub, but the difference is marginal (~5%), and this uses less memory...
    no strict 'refs';
    *{$accessor} = sub {
	(@_ > 1) ? $_[0]->{$property} = $_[1] : $_[0]->{$property};
    }
}

1;

__END__

=head1 DESCRIPTION

The B<accessors::classic> pragma lets you create simple I<classic> Perl
accessors at compile-time.

The generated methods look like this:

  sub foo {
      my $self = shift;
      $self->{foo} = shift if (@_);
      return $self->{foo};
  }

They I<always> return the current value.

Note that there is I<no> dash (C<->) prepended to the property name as there
are in L<accessors>.  This is for backwards compatability.

=head1 PERFORMANCE

There is B<little-to-no performace hit> when using generated accessors; in
fact there is B<usually a performance gain>.

=over 4

=item *

typically I<5-15% faster> than hard-coded accessors (like the above example).

=item *

typically I<1-15% slower> than I<optimized> accessors (less readable).

=item *

typically a I<small> performance hit at startup (accessors are created at
compile-time).

=item *

uses the same anonymous sub to reduce memory consumption (sometimes by 80%).

=back

See the benchmark tests included with this distribution for more details.

=head1 CAVEATS

Classes using blessed scalarrefs, arrayrefs, etc. are not supported for sake
of simplicity.  Only hashrefs are supported.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>

=head1 SEE ALSO

L<accessors>,
L<accessors::rw>,
L<accessors::ro>,
L<accessors::chained>,
L<base>

=cut


lib/accessors.pm
=head1 NAME

accessors - create accessor methods in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors qw( foo bar baz );

  my $obj = bless {}, 'Foo';

  # generates chaining accessors
  # that you can set like this:
  $obj->foo( 'hello ' )
      ->bar( 'world' )
      ->baz( "!\n" );

  # you get the values by passing no params:
  print $obj->foo, $obj->bar, $obj->baz;

=cut

package accessors;

use 5.006;
use strict;
use warnings::register;

our $VERSION  = '1.00';
our $REVISION = (split(/ /, ' $Revision: 1.22 $ '))[2];

our $Debug        = 0;
our $ExportLevel  = 0;
our @InvalidNames = qw( BEGIN CHECK INIT END DESTROY AUTOLOAD );

use constant style => 'chained';

sub import {
    my $class   = shift;
    my $callpkg = caller( $class->ExportLevel );

    my @properties = @_ or return;

    $class->create_accessors_for( $callpkg, @properties );
}

sub create_accessors_for {
    my $class   = shift;
    my $callpkg = shift;

    warn( 'creating ' . $class->style . ' accessors( ',
	  join(' ',@_)," ) in pkg '$callpkg'" ) if $class->Debug;

    foreach my $property (@_) {
	my $accessor = "$callpkg\::$property";
	die( "can't create $accessor - '$property' is not a valid name!" )
	  unless $class->isa_valid_name( $property );
	warn( "creating " . $class->style . " accessor: $accessor\n" ) if
	  $class->Debug > 1;
	$class->create_accessor( $accessor, $property );
    }

    return $class;
}

sub create_accessor {
    my ($class, $accessor, $property) = @_;
    $property = "-$property";
    # set/get is slightly faster if we eval instead of using a closure + anon
    # sub, but the difference is marginal (~5%), and this uses less memory...
    no strict 'refs';
    *{$accessor} = sub {
	(@_ > 1)
	  ? ($_[0]->{$property} = $_[1], return $_[0])
	  : $_[0]->{$property};
    };
}

sub isa_valid_name {
    my ($class, $property) = @_;
    return unless $property =~ /^(?!\d)\w+$/;
    return if grep {$property eq $_} $class->InvalidNames;
    return 1;
}

##
## on the off-chance that someone will sub-class:
##

## don't like studly caps for sub-names, but stick with Exporter-like style...
sub Debug        { $Debug; }
sub ExportLevel  { $ExportLevel }
sub InvalidNames { @InvalidNames }

1;

__END__

=head1 DESCRIPTION

The B<accessors> pragma lets you create simple accessors at compile-time.

This saves you from writing them by hand, which tends to result in
I<cut-n-paste> errors and a mess of duplicated code.  It can also help you
reduce the ammount of unwanted I<direct-variable access> that may creep into
your codebase when you're feeling lazy.  B<accessors> was designed with
laziness in mind.

Method-chaining accessors are generated by default.  Note that you can still
use L<accessors::chained> directly for reasons of backwards compatability.

See L<accessors::classic> for accessors that always return the current value if
you don't like method chaining.

=head1 GENERATED METHODS

B<accessors> will generate methods that return the current object on set:

  sub foo {
      my $self = shift;
      if (@_) { $self->{-foo} = shift; return $self; }
      else    { return $self->{-foo}; }
  }

This way they can be I<chained> together.

=head2 Why prepend the dash?

The dash (C<->) is prepended to the property name for a few reasons:

=over 4

=item *

interoperability with L<Error>.

=item *

to make it difficult to accidentally access the property directly ala:

  use accessors qw( foo );
  $obj->{foo};  # prevents this by mistake
  $obj->foo;    # when you probably meant this

(this might sound woolly, but it's easy enough to do).

=item *

syntactic sugar (this I<is> woolly :).

=back

You shouldn't care too much about how the property is stored anyway - if you do,
you're likely trying to do something special (and should really consider writing
the accessors out long hand), or it's simply a matter of preference in which
case you can use L<accessors::classic>, or sub-class this module.

=head1 PERFORMANCE

There is B<little-to-no performace hit> when using generated accessors; in
fact there is B<usually a performance gain>.

=over 4

=item *

typically I<10-30% faster> than hard-coded accessors (like the above example).

=item *

typically I<1-15% slower> than I<optimized> accessors (less readable).

=item *

typically a I<small> performance hit at startup (accessors are created at
compile-time).

=item *

uses the same anonymous sub to reduce memory consumption (sometimes by 80%).

=back

See the benchmark tests included with this distribution for more details.

=head1 MOTIVATION

The main difference between the B<accessors> pragma and other accessor
generators is B<simplicity>.

=over 4

=item * interface

B<use accessors qw( ... )> is as easy as it gets.

=item * a pragma

it fits in nicely with the B<base> pragma:

  use base      qw( Some::Class );
  use accessors qw( foo bar baz );

and accessors get created at compile-time.

=item * no bells and whistles

The module is extensible instead.

=back

=head1 SUB-CLASSING

If you prefer a different style of accessor or you need to do something more
complicated, there's nothing to stop you from sub-classing.  It should be
pretty easy.  Look through L<accessors::classic>, L<accessors::ro>, and
L<accessors::rw> to see how it's done.

=head1 CAVEATS

Classes using blessed scalarrefs, arrayrefs, etc. are not supported for sake
of simplicity.  Only hashrefs are supported.

=head1 THANKS

Thanks to Michael G. Schwern for indirectly inspiring this module, and for his
feedback & suggestions.

Also to Paul Makepeace and David Wright for showing me faster accessors, to
chocolateboy for his contributions, the CPAN Testers for their bug reports,
and to James Duncan and people on London.pm for their feedback.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>

=head1 SEE ALSO

L<accessors::classic>, L<accessors::chained>

Similar and related modules:

L<base>,
L<fields>,
L<Class::Accessor>,
L<Class::Struct>,
L<Class::Methodmaker>,
L<Class::Generate>,
L<Class::Class>,
L<Class::Tangram>,
L<Object::Tiny>

=cut
